numpy/core/src/multiarray/ctors.c

```diff
@@ -3619,12 +3640,57 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
         }
         PyArray_DIMS(ret)[0] = i + 1;
 
-        if (((item = index2ptr(ret, i)) == NULL) ||
-                (PyArray_DESCR(ret)->f->setitem(value, item, ret) == -1)) {
-            Py_DECREF(value);
-            goto done;
+        if (ndim == 1) {
+            if (((item = index2ptr(ret, i)) == NULL) ||
+                    (PyArray_DESCR(ret)->f->setitem(value, item, ret) == -1)) {
+                Py_DECREF(value);
+                goto done;
+            }
+        }
+        else {
+            iter2 = PyObject_GetIter(value);
```

Comment:
Is this what we want, or should we be checking for sequence types here also, and not just the first value? I'd almost prefer two separate loops, one for `ndim == 1` and the other for `ndim == 2`. Putting everything into one loop somewhat obfuscates the code.